### ISSUE TITLE
Should we also record the original client request when recording caches/scenarios?

### ISSUE THREAD

**Issue by KrekkieD** [KrekkieD/nocca#45]
We've encountered an issue running the same scenario but on different hosts (i.e. one on test, one on acceptation environment). The keygens that were configured when recording the scenario also included the `host` header, as a result only the recording environment could match the key as the host was different on other environments. 

At this point we're in a tough situation, either we record the scenarios again using different keygens (which is a problem when using expiring data as new testdata needs to be requested to record a scenario), or we manually change the keys in the scenarios (but then need to make sure a newly configured keygen matches our manipulated keys).

If we record the original client request we can make something to run that request through the (or a) keygenerator with a new config, and extract the new key from there, then manipulate the scenario with the newly generated key.

It allows more flexibility, but would increase the scenario size, if that matters.

On a sidenote, I wanted to make a GUI option to parse a httpMessage object through a key generator with a custom config, but that was not as easy as I thought.

**Comment by jhkuperus** [comment]
I see your point. I guess what you need to do is parse, mod and export the scenarios again to strip the Host header out. And then change the keygen config to also leave the Host header out.

Or, somewhat inspired by your GUI sidenote: what if you were able to open up a scenario and re-record just a single request? I.e. perform a single request to replace one in a scenario.

On the other hand, you're suggestion of also keeping the original request 'on file' isn't such a bad idea either. Requests are usually quite small, so they won't bog down the file size that much. During replay you could even choose to not load them in memory. This would easily allow you to change the keygen settins, apply, regenerate the keys and be done with it. Thinking about it, I kinda like it :) (Won't help you with your immediate problem though, but it's definitely a nice feature. I vote 👍🏻

--JH

**Comment by jhkuperus** [comment]
I think I'll build it right into the new scenario feature. Can you live with fixing your current situation manually? 

You could use this in the meantime:

```javascript
function stripOrUpdateHostHeaderFromRequestKeys(r, newHost) {
    _.map(r, function(responseObject) {
        var key = JSON.parse(responseObject.requestKey);
        if (newHost) {
            key.headers.Host = newHost;
        }
        else {
            delete key.headers.Host;
        }
        return JSON.stringify(key);
    });
}
```

**Comment by KrekkieD** [comment]
Included in 0.3.0
